internal/pkg/cli/pipeline_init.go

```diff
@@ -225,6 +265,121 @@ func (o *initPipelineOpts) validateURL(url string) error {
+	repos, err := o.fetchAndParseURLs()
+	if err != nil {
+		return err
+	}
+	if _, ok := repos[url]; !ok {
+		return fmt.Errorf("URL '%s' is not a local git remote", o.repoURL)
+	}
+	o.repoRemoteName = repos[url]
+	return nil
+}
+
+func (o *initPipelineOpts) validateBranch() error {
+	// URL has already been checked to exist and be valid; repoRemoteName already set by validateURLExists.
+	branches, err := o.fetchAndParseBranches(o.repoRemoteName)
+	if err != nil {
+		return err
+	}
+	for _, branch := range branches {
+		if branch == o.repoBranch {
+			return nil
+		}
+	}
+	return fmt.Errorf("branch '%s' not found for repo '%s'", o.repoBranch, o.repoRemoteName)
+}
+
+func (o *initPipelineOpts) fetchAndParseURLs() (map[string]string, error) {
+	// Fetches and parses all remote repositories; returns a map with `url: remoteName`.
+	err := o.runner.Run("git", []string{"remote", "-v"}, exec.Stdout(&o.repoBuffer))
+	if err != nil {
+		return nil, fmt.Errorf("get Git remote repository info: %w", err)
+	}
+	repos := o.parseGitRemoteResult(strings.TrimSpace(o.repoBuffer.String()))
+	o.repoBuffer.Reset()
+
+	return repos, nil
+}
+
+func (o *initPipelineOpts) fetchAndParseBranches(remoteName string) ([]string, error) {
+	// Fetches and parses all branches associated with the chosen repo.
+	err := o.runner.Run("git", []string{"branch", "-a", "-l", remoteName + "/*"}, exec.Stdout(&o.branchBuffer))
+	if err != nil {
+		return nil, fmt.Errorf("get Git repo branch info: %w", err)
+	}
+	if o.branchBuffer.String() == "" {
+		return []string{}, nil
+	}
+	branches, err := o.parseGitBranchResults(strings.TrimSpace(o.branchBuffer.String()))
+	if err != nil {
+		return nil, fmt.Errorf("parse 'git branch' results: %w", err)
+	}
+	o.branchBuffer.Reset()
+	return branches, nil
+}
+
+func (o *initPipelineOpts) parseAndAssignRepoParts(repo string) error {
+	repoParts := strings.Split(repo, ": ")
+	if len(repoParts) != 2 {
+		return fmt.Errorf("cannot parse repo '%s'", repo)
+	}
+	o.repoURL = repoParts[1]
+	o.repoRemoteName = repoParts[0]
+	return nil
+}
+
+func (o *initPipelineOpts) filterThenSelectReposByBranch() error {
+	var reposWithBranch []string
+	URLsAndRemoteNames, err := o.fetchAndParseURLs()
+	if err != nil {
+		return err
+	}
+	for url, remoteName := range URLsAndRemoteNames {
+		branches, err := o.fetchAndParseBranches(remoteName)
+		if err != nil {
+			return err
+		}
+		if len(branches) == 0 {
+			continue
+		}
+		for _, branch := range branches {
+			if branch == o.repoBranch {
+				reposWithBranch = append(reposWithBranch, fmt.Sprintf("%s: %s", remoteName, url))
+			}
+		}
+	}
+
+	if len(reposWithBranch) == 1 {
+		log.Infof("Only one git repository detected. Your pipeline will follow '%s'. You may make changes in the generated pipeline manifest before deployment.\n", color.HighlightUserInput(reposWithBranch[0]))
+		if err := o.parseAndAssignRepoParts(reposWithBranch[0]); err != nil {
+			return nil
+		}
+
```

Comment:
Hmm I'd prefer it to be an isolated function so that we can explicitly know when we assign values to `o.repoURL` and `o.repoRemoteName`.

```suggestion
	 repoURL, repoRemoteName, err := o.parseAndAssignRepoParts(reposWithBranch[0])
	 if err != nil {
			return nil
	}
	o.repoURL, o.repoRemoteName = repoURL, repoRemoteName
```